core/src/main/scala/ox/OxApp.scala

```
@@ -0,0 +1,65 @@
+package ox
+
+import scala.util.boundary.*
+import scala.util.control.NonFatal
+
+enum ExitCode(val code: Int):
+  case Success extends ExitCode(0)
+  case Failure(exitCode: Int = 1) extends ExitCode(exitCode)
+
+trait OxApp:
+  def main(args: Array[String]): Unit =
+    unsupervised:
+      val cancellableMainFork = forkCancellable(supervised(handleRun(args.toVector)))
+
+      val interruptThread = new Thread(() => {
+        cancellableMainFork.cancel()
+        ()
+      })
+
+      interruptThread.setName("ox-interrupt-hook")
+
+      mountShutdownHook(interruptThread)
+
+      cancellableMainFork.joinEither() match
+        case Left(iex: InterruptedException) => exit(0)
+        case Left(fatalErr)                  => throw fatalErr
+        case Right(exitCode)                 => exit(exitCode.code)
+
+  private[ox] def exit(code: Int): Unit = System.exit(code)
+
+  private[ox] def mountShutdownHook(thread: Thread): Unit =
+    try Runtime.getRuntime.addShutdownHook(thread)
+    catch case _: IllegalStateException => ()
```

shouldn't we re-throw the exception if mounding the hook fails?

---

for reference: CE IOApp only shuts down it's thread pools when it gets ISE here